modules/drivers/presto-jdbc/src/metabase/driver/presto_jdbc.clj

```diff
@@ -0,0 +1,489 @@
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defmethod sql-jdbc.execute/prepared-statement :presto-jdbc
+  [driver ^Connection conn ^String sql params]
+  ;; with Presto JDBC driver, result set holdability must be HOLD_CURSORS_OVER_COMMIT
+  ;; defining this method simply to omit setting the holdability
+  (let [stmt (.prepareStatement conn
+                                sql
+                                ResultSet/TYPE_FORWARD_ONLY
+                                ResultSet/CONCUR_READ_ONLY)]
+       (try
+         (try
+           (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+           (catch Throwable e
+             (log/debug e (trs "Error setting prepared statement fetch direction to FETCH_FORWARD"))))
+         (sql-jdbc.execute/set-parameters! driver stmt params)
+         stmt
+         (catch Throwable e
+           (.close stmt)
+           (throw e)))))
+
+
+(defmethod sql-jdbc.execute/statement :presto-jdbc
+  [_ ^Connection conn]
+  ;; and similarly for statement (do not set holdability)
+  (let [stmt (.createStatement conn
+                               ResultSet/TYPE_FORWARD_ONLY
+                               ResultSet/CONCUR_READ_ONLY)]
+    (try
+      (.setFetchDirection stmt ResultSet/FETCH_FORWARD)
+      (catch Throwable e
+        (log/debug e (trs "Error setting statement fetch direction to FETCH_FORWARD"))))
+    stmt))
+
+(defmethod driver/can-connect? :sql-jdbc
+  [driver details]
+  (sql-jdbc.conn/can-connect? driver details))
+
+(defn- ^PrestoConnection pooled-conn->presto-conn
+  "Unwraps the C3P0 `pooled-conn` and returns the underlying `PrestoConnection` it holds."
+  [^C3P0ProxyConnection pooled-conn]
+  (.unwrap pooled-conn PrestoConnection))
+
+(defmethod sql-jdbc.execute/connection-with-timezone :presto-jdbc
+  [driver database ^String timezone-id]
+  ;; Presto supports setting the session timezone via a `PrestoConnection` instance method. Under the covers,
+  ;; this is equivalent to the `X-Presto-Time-Zone` header in the HTTP request (i.e. the `:presto` driver)
+  (let [conn            (.getConnection (sql-jdbc.execute/datasource database))
+        underlying-conn (pooled-conn->presto-conn conn)]
+    (try
+      (sql-jdbc.execute/set-best-transaction-level! driver conn)
+      (when-not (str/blank? timezone-id)
+        ;; set session time zone if defined
+        (.setTimeZoneId underlying-conn timezone-id))
+      (try
+        (.setReadOnly conn true)
+        (catch Throwable e
+          (log/debug e (trs "Error setting connection to read-only"))))
+      ;; as with statement and prepared-statement, cannot set holdability on the connection level
+      conn
+      (catch Throwable e
+        (.close conn)
+        (throw e)))))
+
+#_(defmethod sql-jdbc.execute/set-parameter [:presto-jdbc OffsetDateTime]
```

Comment:
just confirming, are these meant to be commented out?

Reply:
I had a bit of back and forth with Cam on this, on a different thread (don't know how to link directly to other GitHub PR comment threads).  But in essence, this is theoretically not needed anymore because I have overridden the `->prepared-substitution` method for all of these temporal types, which then go on to set the "values" as strings.  However, if there are any other code paths that end up calling `set-parameter`, then those will definitely fail with an error, because the Presto JDBC driver doesn't support the override that ends up being called. 

The fact that they're commented out, but the tests pass, suggests strongly that no such code path exists, but I was considering adding them back in just out of sheer paranoia.  I think for now, I will just remove and then we can add them if we have an identified bug in the future.

Reply:
cool. yeah i vote remove them as well